File: leetcode_java/src/main/java/LeetCodeJava/Array/MaximumSumOfTwoNonOverlappingSubarrays.java
package LeetCodeJava.Array;

// https://leetcode.com/problems/maximum-sum-of-two-non-overlapping-subarrays/description/

/**
 * 1031. Maximum Sum of Two Non-Overlapping Subarrays
 * Medium
 * Topics
 * Companies
 * Hint
 * Given an integer array nums and two integers firstLen and secondLen, return the maximum sum of elements in two non-overlapping subarrays with lengths firstLen and secondLen.
 * <p>
 * The array with length firstLen could occur before or after the array with length secondLen, but they have to be non-overlapping.
 * <p>
 * A subarray is a contiguous part of an array.
 * <p>
 * <p>
 * <p>
 * Example 1:
 * <p>
 * Input: nums = [0,6,5,2,2,5,1,9,4], firstLen = 1, secondLen = 2
 * Output: 20
 * Explanation: One choice of subarrays is [9] with length 1, and [6,5] with length 2.
 * Example 2:
 * <p>
 * Input: nums = [3,8,1,3,2,1,8,9,0], firstLen = 3, secondLen = 2
 * Output: 29
 * Explanation: One choice of subarrays is [3,8,1] with length 3, and [8,9] with length 2.
 * Example 3:
 * <p>
 * Input: nums = [2,1,5,6,0,9,5,0,3,8], firstLen = 4, secondLen = 3
 * Output: 31
 * Explanation: One choice of subarrays is [5,6,0,9] with length 4, and [0,3,8] with length 3.
 * <p>
 * <p>
 * Constraints:
 * <p>
 * 1 <= firstLen, secondLen <= 1000
 * 2 <= firstLen + secondLen <= 1000
 * firstLen + secondLen <= nums.length <= 1000
 * 0 <= nums[i] <= 1000
 */
public class MaximumSumOfTwoNonOverlappingSubarrays {

    // V0
    //    public int maxSumTwoNoOverlap(int[] nums, int firstLen, int secondLen) {
    //
    //    }


    // V-0-1
    // IDEA: Prefix Sum + Sliding Window (GPT)
    public int maxSumTwoNoOverlap_0_1(int[] nums, int firstLen, int secondLen) {
        int n = nums.length;

        // build prefix sum
        int[] prefix = new int[n + 1];
        for (int i = 0; i < n; i++) {
            prefix[i + 1] = prefix[i] + nums[i];
        }

        /** NOTE !!!
         *
         * We do both:
         *
         * firstLen → secondLen
         * secondLen → firstLen
         *
         */
        return Math.max(
                helper(prefix, firstLen, secondLen),
                helper(prefix, secondLen, firstLen)
        );
    }

    /** NOTE !!!
     *
     *  L, M are `len` of sub array
     *    - firstLen, secondLen
     *    - secondLen, firstLen
     *
     * L: length of the first subarray (must come first)
     * M: length of the second subarray (comes after L)
     *
     */
    // L comes before M
    private int helper(int[] prefix, int L, int M) {
        // maxL: the maximum sum of
        // `any L-length` subarray seen so far
        // (on the left side)
        int maxL = 0;
        int res = 0;

        /**  NOTE !!!
         *
         * Why start at L + M?
         *
         * ->
         * We need:
         *   an L-length subarray
         *   AND an M-length subarray after it
         *    -> So minimum total length = L + M
         *
         * At each i, we treat:
         *
         *    ->
         *
         *    M-subarray ends at index i-1
         *    L-subarray must end before that
         *
         */
        for (int i = L + M; i < prefix.length; i++) {

            /** NOTE !!!
             *
             *  1.
             * maxL:
             *   keeps the `best` `L-length` subarray
             *   before current position
             *
             *
             * 2.
             * prefix[i - M] - prefix[i - M - L]
             *  -> sum of subarray of length L ending at index (i - M - 1)
             *  -> why ?
             *     - i - M → start of current M window
             *       -> So L must end before that
             *       -> The L window is:
             *           ```
             *           [i - M - L, ..., i - M - 1]
             *           ```
             *
             * 3.
             *
             * So this line does:
             *
             *    Compute current candidate L-window
             *    Compare with previous best
             *    Store the maximum in maxL
             *
             *
             *
             */
            // best L subarray before current M
            maxL = Math.max(maxL, prefix[i - M] - prefix[i - M - L]);


            /** NOTE !!!
             *
             * sum of subarray of length M ending at index i-1
             *
             */
            // current M subarray
            int currM = prefix[i] - prefix[i - M];

            /**  NOTE !!
             *
             * Now we combine:
             *
             *   - best L before
             *   - current M
             *
             *
             * ->
             *
             * This guarantees:
             * ✅ no overlap
             * ✅ optimal pairing so far
             *
             *
             */
            res = Math.max(res, maxL + currM);
        }

        return res;
    }



    // V0-2
    // IDEA: Prefix Sum + Sliding Window  (gemini)
    public int maxSumTwoNoOverlap_0_2(int[] nums, int firstLen, int secondLen) {
        int n = nums.length;
        int[] prefixSum = new int[n + 1];
        for (int i = 0; i < n; i++) {
            prefixSum[i + 1] = prefixSum[i] + nums[i];
        }

        // We need to check both cases:
        // 1. firstLen before secondLen
        // 2. secondLen before firstLen
        return Math.max(solve(prefixSum, firstLen, secondLen),
                solve(prefixSum, secondLen, firstLen));
    }

    private int solve(int[] p, int L, int M) {
        int maxL = 0;
        int res = 0;

        // We start 'i' at a point where at least one L-length
        // and one M-length subarray can exist.
        for (int i = L + M; i < p.length; i++) {

            // p[i-M] - p[i-M-L] is the L-length
            // window ending just before the M-window starts
            maxL = Math.max(maxL, p[i - M] - p[i - M - L]);

            // p[i] - p[i-M] is the current M-length window
            res = Math.max(res, maxL + (p[i] - p[i - M]));
        }

        return res;
    }


    // V0-4
    // IDEA: PREFIX SUM
    // time: O(N), space: O(N)

    /**
     * Does This Cover All Configurations?
     * <p>
     * Yes, it does. By handling the two cases separately, the algorithm ensures:
     * •	All valid configurations are checked:
     * •	firstLen subarray is before secondLen.
     * •	secondLen subarray is before firstLen.
     * •	There’s no need to explicitly consider cases where:
     * •	secondLen subarray comes before firstLen in the “case 1” loop because “case 2” handles that.
     * •	Similarly, the converse is true for “case 2”.
     */
    public int maxSumTwoNoOverlap_0_4(int[] nums, int firstLen, int secondLen) {
        int n = nums.length;
        int[] s = new int[n + 1];
        for (int i = 0; i < n; ++i) {
            s[i + 1] = s[i] + nums[i];
        }
        int ans = 0;

        /**
         * where end is exclusive and start is inclusive.
         * 	2.	Two Cases:
         * 	•	Case 1: Calculate the maximum sum when a subarray of length firstLen
         * 	            comes before a subarray of length secondLen.
         *
         * 	•	Case 2: Calculate the maximum sum when a subarray of length secondLen
         * 	            comes before a subarray of length firstLen.
         *
         *
         */

        // case 1) check `firstLen`, then `secondLen`
        for (int i = firstLen, t = 0; i + secondLen - 1 < n; ++i) {
            /**
             *  - Logic:
             *
             * 	1.	Start at index i = firstLen because the first subarray
             *       	must be at least firstLen long.
             *
             * 	2.	Use t to track the maximum sum of a subarray of length firstLen
             *    	ending at or before index i.
             *
             * 	3.	Add the sum of a subarray of length secondLen starting at index i
             *    	and update ans with the combined sum.
             *
             *
             *    NOTE !!!
             *
             *
             *  - In this case:
             *    - The loop ensures the first subarray (firstLen) is placed before the second subarray (secondLen).
             * 	- The variable t keeps track of the maximum sum of the firstLen subarray that ends before the second subarray starts at index i.
             *  - Important Note:
             *    - The secondLen subarray must always start after the firstLen subarray ends because the iteration proceeds sequentially.
             */
            t = Math.max(t, s[i] - s[i - firstLen]); // Maximum sum of `firstLen` ending at or before `i`
            ans = Math.max(ans, t + s[i + secondLen] - s[i]); // Combine with `secondLen` starting at `i`
        }
        // case 2) check `secondLen`, then `firstLen`

        /**
         *  - Logic
         *
         *      * 	1.	Start at index i = secondLen because the first subarray must be at least secondLen long.
         *      * 	2.	Use t to track the maximum sum of a subarray of length secondLen ending at or before index i.
         *      * 	3.	Add the sum of a subarray of length firstLen starting at index i and update ans with the combined sum.
         *
         *  - Goal: Calculate the maximum sum where a subarray of length secondLen comes before a subarray of length firstLen.
         *
         *  NOTE !!!
         *
         * - In this case:
         *      - The loop ensures the first subarray (secondLen) is placed before the second subarray (firstLen).
         *      - The variable t keeps track of the maximum sum of the secondLen subarray that ends before the first subarray starts at index i.
         * 	- Important Note:
         *      - The firstLen subarray must always start after the secondLen subarray ends because the iteration proceeds sequentially.
         *
         */
        for (int i = secondLen, t = 0; i + firstLen - 1 < n; ++i) {
            t = Math.max(t, s[i] - s[i - secondLen]);
            ans = Math.max(ans, t + s[i + firstLen] - s[i]);
        }
        return ans;
    }


    // V1
    // https://leetcode.ca/2018-09-26-1031-Maximum-Sum-of-Two-Non-Overlapping-Subarrays/
    // IDEA: PREFIX SUM

    /**
     * time = O(N)
     * space = O(N)
     */
    public int maxSumTwoNoOverlap_1(int[] nums, int firstLen, int secondLen) {
        int n = nums.length;
        int[] s = new int[n + 1];
        for (int i = 0; i < n; ++i) {
            s[i + 1] = s[i] + nums[i];
        }
        int ans = 0;
        // case 1)  check `firstLen`, then `secondLen`
        for (int i = firstLen, t = 0; i + secondLen - 1 < n; ++i) {
            t = Math.max(t, s[i] - s[i - firstLen]);
            ans = Math.max(ans, t + s[i + secondLen] - s[i]);
        }
        // case 2)  check  `secondLen`, then `firstLen`
        for (int i = secondLen, t = 0; i + firstLen - 1 < n; ++i) {
            t = Math.max(t, s[i] - s[i - secondLen]);
            ans = Math.max(ans, t + s[i + firstLen] - s[i]);
        }
        return ans;
    }

    // V1-2
    // IDEA: PREFIX SUM
    // https://leetcode.com/problems/maximum-sum-of-two-non-overlapping-subarrays/solutions/1489581/java-easy-to-understand-prefix-sum-by-rm-5d2z/

    /**
     * time = O(N)
     * space = O(N)
     */
    public int maxSumTwoNoOverlap_1_2(int[] A, int L, int M) {
        int sums[] = new int[A.length + 1];

        for (int i = 1; i <= A.length; i++)
            sums[i] = A[i - 1] + sums[i - 1];

        int maxLval = 0;
        int ans = 0;
        for (int i = L; i <= A.length - M; i++) {
            maxLval = Math.max(maxLval, sums[i] - sums[i - L]);
            ans = Math.max(ans, sums[i + M] - sums[i] + maxLval);
        }
        int maxRval = 0;
        for (int i = M; i <= A.length - L; i++) {
            maxRval = Math.max(maxRval, sums[i] - sums[i - M]);
            ans = Math.max(ans, sums[i + L] - sums[i] + maxRval);
        }
        return ans;
    }

    // V2
    // IDEA: PREFIX SUM
    // https://leetcode.com/problems/maximum-sum-of-two-non-overlapping-subarrays/solutions/5171249/java-clean-solution-using-prefix-sum-det-lxok/
    // time: O(N), space: O(N)
    // Drive Function
    public int maxSumTwoNoOverlap_2(int[] nums, int firstLen, int secondLen) {
        int[] prefix = new int[nums.length + 1];
        for (int i = 0; i < nums.length; i++) {
            prefix[i + 1] = prefix[i] + nums[i];
        }

        return Math.max(solveHelper(prefix, firstLen, secondLen),
                solveHelper(prefix, secondLen, firstLen));
    }

    private int solveHelper(int[] prefix, int len1, int len2) {
        int res = 0;
        int maxLength = Integer.MIN_VALUE;
        for (int i = len1 + len2; i < prefix.length; i++) {
            maxLength = Math.max(maxLength,
                    prefix[i - len2] - prefix[i - len2 - len1]);
            res = Math.max(res, maxLength + prefix[i] - prefix[i - len2]);
        }
        return res;
    }

    // V3
    // https://leetcode.com/problems/maximum-sum-of-two-non-overlapping-subarrays/solutions/913234/on-thought-process-java-by-133c7-wn3a/

    /**
     * time = O(N)
     * space = O(N)
     */
    public int maxSumTwoNoOverlap_3(int[] A, int L, int M) {
        int n = A.length;
        if (n == 0) {
            return 0;
        }
        if (L + M > n) {
            throw new IllegalArgumentException();
        }

        int[] Ls = new int[n];
        int[] Ms = new int[n];

        int lSum = 0;
        int mSum = 0;

        for (int i = 0; i <= n - L; i++) {
            if (i == 0) {
                for (int j = 0; j < L; j++) {
                    lSum += A[j];
                }
            } else {
                lSum = lSum - A[i - 1] + A[i + L - 1];
            }

            Ls[i] = lSum;
        }

        for (int i = 0; i <= n - M; i++) {
            if (i == 0) {
                for (int j = 0; j < M; j++) {
                    mSum += A[j];
                }
            } else {
                mSum = mSum - A[i - 1] + A[i + M - 1];
            }

            Ms[i] = mSum;
        }

        int[] lmax = new int[n + 1];
        int[] mmax = new int[n + 1];

        for (int i = n - 1; i >= 0; i--) {
            lmax[i] = Math.max(lmax[i + 1], Ls[i]);
            mmax[i] = Math.max(mmax[i + 1], Ms[i]);
        }

        int ans = 0;
        for (int i = 0; i <= n - L - 1; i++) {
            ans = Math.max(ans, Ls[i] + mmax[i + L]);
        }
        for (int i = 0; i <= n - M - 1; i++) {
            ans = Math.max(ans, Ms[i] + lmax[i + M]);
        }

        return ans;
    }





}

File: leetcode_python/Array/maximum-sum-of-two-non-overlapping-subarrays.py
# https://leetcode.com/problems/maximum-sum-of-two-non-overlapping-subarrays/description/

"""

1031. Maximum Sum of Two Non-Overlapping Subarrays
Solved
Medium
Topics
premium lock icon
Companies
Hint
Given an integer array nums and two integers firstLen and secondLen, return the maximum sum of elements in two non-overlapping subarrays with lengths firstLen and secondLen.

The array with length firstLen could occur before or after the array with length secondLen, but they have to be non-overlapping.

A subarray is a contiguous part of an array.

 

Example 1:

Input: nums = [0,6,5,2,2,5,1,9,4], firstLen = 1, secondLen = 2
Output: 20
Explanation: One choice of subarrays is [9] with length 1, and [6,5] with length 2.
Example 2:

Input: nums = [3,8,1,3,2,1,8,9,0], firstLen = 3, secondLen = 2
Output: 29
Explanation: One choice of subarrays is [3,8,1] with length 3, and [8,9] with length 2.
Example 3:

Input: nums = [2,1,5,6,0,9,5,0,3,8], firstLen = 4, secondLen = 3
Output: 31
Explanation: One choice of subarrays is [5,6,0,9] with length 4, and [0,3,8] with length 3.
 

Constraints:

1 <= firstLen, secondLen <= 1000
2 <= firstLen + secondLen <= 1000
firstLen + secondLen <= nums.length <= 1000
0 <= nums[i] <= 1000


"""


# V0
# IDEA: PREFIX SUM (gpt)
"""

Core idea:


-> `precompute prefix sums`, then scan the array `twice`:

L before M: the firstLen subarray comes 
            before the secondLen subarray.

M before L: the reverse order.

During each scan, maintain the maximum sum 
of the left subarray seen so far,
then combine it with the current right subarray.


"""
class Solution(object):
    def maxSumTwoNoOverlap(self, nums, firstLen, secondLen):
        n = len(nums)

        """
        NOTE !!!


        prefix size = `n+1`
        """
        prefix = [0] * (n + 1)
        for i in range(n):
            """
            NOTE !!!


            how we prepare prefix array:

            https://yennj12.js.org/CS_basics/cheatsheets/prefix_sum.html


            -> prefix[i+1] = prefix[i] + nums[i]
            """
            prefix[i + 1] = prefix[i] + nums[i]

        return max(
            self.helper(nums, prefix, firstLen, secondLen),
            self.helper(nums, prefix, secondLen, firstLen)
        )

    
    def helper(self, nums, prefix, firstLen, secondLen):
        n = len(nums)

        # NOTE !!!
        # we get cur `bestFirst` first
        bestFirst = prefix[firstLen] - prefix[0]
        ans = 0

        """
        NOTE !!!

        
        1. 
          ONLY 1 loop:
            range(firstLen + secondLen, n + 1)


        2. we get updated  `bestFirst`

        3. we get second

        4. we get cur max ans
        """
        for i in range(firstLen + secondLen, n + 1):
            bestFirst = max(
                bestFirst,
                prefix[i - secondLen] - prefix[i - secondLen - firstLen]
            )

            second = prefix[i] - prefix[i - secondLen]

            ans = max(ans, bestFirst + second)

        return ans



# V0-0-1
# IDEA: PREFIX SUM (gemini)
class Solution(object):
    def maxSumTwoNoOverlap(self, nums, firstLen, secondLen):
        """
        :type nums: List[int]
        :type firstLen: int
        :type secondLen: int
        :rtype: int
        """
        n = len(nums)
        
        # Build an N+1 sized prefix sum array.
        # prefix[i] will store the sum of all elements before index i.
        prefix = [0] * (n + 1)
        for i in range(n):
            prefix[i + 1] = prefix[i] + nums[i]
            
        # Scenario 1: firstLen comes BEFORE secondLen
        v1 = self.helper(prefix, n, firstLen, secondLen)
        
        # Scenario 2: secondLen comes BEFORE firstLen
        v2 = self.helper(prefix, n, secondLen, firstLen)
        
        return max(v1, v2)
        

    # NOTE !!!!
    # below helper using `double loop`
    def helper(self, prefix, n, len1, len2):
        max_sum = 0
        
        # i is the start index of the first window
        # It can go up to n - len1 - len2 (to leave room for the second window)
        for i in range(n - len1 - len2 + 1):
            
            # The sum of nums[i : i + len1] using our prefix array
            sum1 = prefix[i + len1] - prefix[i]
            
            # j is the start index of the second window
            # It must start strictly after the first window ends (i + len1)
            for j in range(i + len1, n - len2 + 1):
                
                # The sum of nums[j : j + len2]
                sum2 = prefix[j + len2] - prefix[j]
                
                # Keep track of the maximum combined sum
                max_sum = max(max_sum, sum1 + sum2)
                
        return max_sum


# V0
# IDEA: BRUTE FORCE + `Try both orders` (gpt)
# https://github.com/yennanliu/CS_basics/blob/master/leetcode_java/src/main/java/LeetCodeJava/Array/MaximumSumOfTwoNonOverlappingSubarrays.java#L50
"""
CORE IDEA:


 ```
 Try both orders:

 1. firstLen before secondLen
 2. secondLen before firstLen

 For each order:
 - maintain best L-window seen so far
 - combine with current M-window
 ```
"""
class Solution(object):
    def maxSumTwoNoOverlap(self, nums, firstLen, secondLen):
        n = len(nums)
        max_sum = 0

        # firstLen before secondLen
        for i in range(n - firstLen + 1):
            f_window = nums[i:i + firstLen]

            for j in range(i + firstLen, n - secondLen + 1):
                s_window = nums[j:j + secondLen]
                max_sum = max(max_sum, sum(f_window) + sum(s_window))

        # secondLen before firstLen
        for i in range(n - secondLen + 1):
            f_window = nums[i:i + secondLen]

            for j in range(i + secondLen, n - firstLen + 1):
                s_window = nums[j:j + firstLen]
                max_sum = max(max_sum, sum(f_window) + sum(s_window))

        return max_sum


# V1-1
# IDEA: PREFIX SUM (gpt)
"""

Core idea:


-> `precompute prefix sums`, then scan the array `twice`:

L before M: the firstLen subarray comes 
            before the secondLen subarray.

M before L: the reverse order.

During each scan, maintain the maximum sum 
of the left subarray seen so far,
then combine it with the current right subarray.


"""
class Solution:
    def maxSumTwoNoOverlap(self, nums, firstLen, secondLen):
        n = len(nums)

        # prefix[i] = sum(nums[:i])
        prefix = [0] * (n + 1)
        for i in range(n):
            prefix[i + 1] = prefix[i] + nums[i]

        def maxSum(L, M):
            # Best L-length window before current M window
            bestL = prefix[L] - prefix[0]
            ans = 0

            # i = starting index of M window
            for i in range(L, n - M + 1):
                # Update best L window ending before i
                bestL = max(bestL, prefix[i] - prefix[i - L])

                # Current M window
                currM = prefix[i + M] - prefix[i]

                ans = max(ans, bestL + currM)

            return ans

        return max(maxSum(firstLen, secondLen),
                   maxSum(secondLen, firstLen))



# V1-2
# IDEA: PREFIX SUM (gemini)
class Solution(object):
    def maxSumTwoNoOverlap(self, nums, firstLen, secondLen):
        """
        :type nums: List[int]
        :type firstLen: int
        :type secondLen: int
        :rtype: int
        """
        n = len(nums)
        
        # 1. Build a prefix sum array to instantly calculate subarray sums
        # prefix[i] represents the sum of nums[0] to nums[i-1]
        prefix = [0] * (n + 1)
        for i in range(n):
            prefix[i + 1] = prefix[i] + nums[i]
            
        max_L = 0  # Maximum sum of a firstLen subarray seen so far
        max_M = 0  # Maximum sum of a secondLen subarray seen so far
        max_total = 0
        
        L = firstLen
        M = secondLen
        
        # 2. Iterate starting from the point where BOTH subarrays can fit (L + M)
        for i in range(L + M, n + 1):
            
            # --- Scenario A: L comes before M ---
            # Sum of the L-length subarray that ends exactly where the current M-length subarray starts
            curr_L = prefix[i - M] - prefix[i - M - L]
            max_L = max(max_L, curr_L)
            
            # Sum of the current M-length subarray ending at index i
            sum_M = prefix[i] - prefix[i - M]
            
            # --- Scenario B: M comes before L ---
            # Sum of the M-length subarray that ends exactly where the current L-length subarray starts
            curr_M = prefix[i - L] - prefix[i - L - M]
            max_M = max(max_M, curr_M)
            
            # Sum of the current L-length subarray ending at index i
            sum_L = prefix[i] - prefix[i - L]
            
            # Update the global maximum using the best combination from either scenario
            max_total = max(max_total, max_L + sum_M, max_M + sum_L)
            
        return max_total



# V1-3
# IDEA: PREFIX SUM (gpt)
class Solution:
    def maxSumTwoNoOverlap(self, nums, firstLen, secondLen):
        n = len(nums)
        pre = [0]
        for x in nums:
            pre.append(pre[-1] + x)

        def solve(L, M):
            best = 0
            ans = 0
            for i in range(L, n - M + 1):
                best = max(best, pre[i] - pre[i - L])
                ans = max(ans, best + pre[i + M] - pre[i])
            return ans

        return max(solve(firstLen, secondLen),
                   solve(secondLen, firstLen))



# V2
# https://leetcode.com/problems/maximum-sum-of-two-non-overlapping-subarrays/solutions/6575097/maximum-sum-of-two-non-overlapping-subar-abkc/
class Solution(object):
    def maxSumTwoNoOverlap(self, nums, firstLen, secondLen):
        """
        :type nums: List[int]
        :type firstLen: int
        :type secondLen: int
        :rtype: int
        """
        def getMaxSum(first, second):
            maxFirst, res = 0, 0
            prefix = [0] * (len(nums) + 1)  

            for i in range(len(nums)):
                prefix[i + 1] = prefix[i] + nums[i]

            for i in range(first + second, len(nums) + 1):
                maxFirst = max(maxFirst, prefix[i - second] - prefix[i - first - second])
                res = max(res, maxFirst + prefix[i] - prefix[i - second])
            return res

        return max(getMaxSum(firstLen, secondLen), getMaxSum(secondLen, firstLen))


# V3
# https://leetcode.com/problems/maximum-sum-of-two-non-overlapping-subarrays/solutions/760194/python-clean-on-98-by-warmr0bot-9k9z/
class Solution:
    def maxSumTwoNoOverlap(self, A: List[int], L: int, M: int) -> int:
        prefix = [0]
        maxl = maxm = summ = 0
        for x in A:
            prefix.append(prefix[-1] + x)
        
        for x in range(M, len(prefix) - L):
            maxm = max(maxm, prefix[x] - prefix[x - M])
            summ = max(summ, maxm + prefix[x + L] - prefix[x])
        
        for x in range(L, len(prefix) - M):
            maxl = max(maxl, prefix[x] - prefix[x - L])
            summ = max(summ, maxl + prefix[x + M] - prefix[x])
        
        return summ
